File: packages/native_pdf_renderer/android/src/main/kotlin/io/scer/native_pdf_renderer/resources/Repository.kt
package io.scer.native_pdf_renderer.resources

import java.lang.Exception

class RepositoryItemNotFoundException(message: String): Exception(message)

abstract class Repository<T> {
    private val items: MutableMap<String, T> = HashMap()

    @Throws(RepositoryItemNotFoundException::class)
    fun get(id: String): T {
        if (!exist(id)) throw RepositoryItemNotFoundException(id)
        return items[id]!!
    }

    fun set(id: String, item: T) {
        items[id] = item
    }

    private fun exist(id: String): Boolean = items.contains(id)

    protected open fun close(id: String) {
        items.remove(id)
    }
}

File: packages/pdfx/lib/src/renderer/web/resources/repository.dart
import 'package:meta/meta.dart';
import 'package:uuid/uuid.dart';

const Uuid uuid = Uuid();

abstract class Repository<T> {
  final _items = <String?, T>{};

  T? get(String? id) {
    if (!_exist(id)) {
      throw RepositoryItemNotFoundException();
    }
    return _items[id];
  }

  void set(String? id, T item) {
    _items[id] = item;
  }

  bool _exist(String? id) => _items.containsKey(id);

  @protected
  void close(String? id) {
    _items.remove(id);
  }
}

class RepositoryItemNotFoundException implements Exception {}
